### ISSUE TITLE
[ENH] Have a way to disable construction from Python

### ISSUE THREAD

**Issue by da-woods** [cython/cython#5319]
### Is your feature request related to a problem? Please describe.

It's a fairly common [and documented](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html?highlight=staticmethod#instantiation-from-existing-c-c-pointers) pattern to use cdef `@staticmethod`s to construct a class from C types that cannot be converted to/from Python.

However, this doesn't stop users from constructing the type normally from within Python (which may put it in an invalid state). The example in the documentation kind of does this by raising a `TypeError` from `__init__`, but it's possible to get around it by calling `__new__` from Python. It also makes instantiating it from the `staticmethod` somewhat fiddly.

### Describe the solution you'd like.

It'd be nice to have a decorator to say "this type can only be constructed from Cython". `@internal` kind of does that, but isn't quite the same and isn't really documented (`@internal` hides the class, but once you have access to an instance of it I think you can still construct it).

### Additional context

There's some fiddly decisions about inheritance, and what we allow/disallow in terms of `__cinit__`.

**Comment by da-woods** [comment]
>  Can this be prevented by raising an error in `__new__`?

no, because the static method has to use `__new__` to construct the class (which doesn't necessarily have to be the case - it's possible to skip `__new__` in the C API, but Cython doesn't support it.

It clearly isn't an essential feature because the status quo does work - it'd just be a case of letting people control construction a little more.

**Comment by scoder** [comment]
I agree with "we're all adults". Being able to crash the Python interpreter when you're already able to execute arbitrary code is rarely a real threat. Being able to hide internal types from the module API seems helpful, but adding safe-guards against malicious instantiation seems overkill. If it's not easy to do, it's probably not worth it.
